packages/contracts/contracts/web3call/PMPV0.sol

```diff
@@ -0,0 +1,979 @@
+
+    bytes32 private constant _TYPE = "PMPV0";
+
+    bytes16 private constant _HEX_DIGITS = "0123456789abcdef";
+    uint256 private constant _DECIMAL_PRECISION_DIGITS = 10;
+    uint256 private constant _DECIMAL_PRECISION =
+        10 ** _DECIMAL_PRECISION_DIGITS;
+
+    // @dev min hex color assumed to be 0x000000
+    uint256 private constant _HEX_COLOR_MAX = 0xFFFFFF;
+
+    uint256 private constant _TIMESTAMP_MIN = 0; // @dev unix timestamp, 0 = 1970-01-01
+    uint256 private constant _TIMESTAMP_MAX = type(uint64).max; // max guardrail, ~10 billion years
+
+    // @dev core contract address and projectId are implicit based on mapping pointing to ProjectConfig struct
+    struct ProjectConfig {
+        // @dev array of pmpKeys for efficient enumeration, uses efficient SSTORE2 storage
+        ImmutableStringArray.StringArray pmpKeys; // slot 0: 32 bytes
+        // @dev mapping of pmpKeys to PMPConfigStorage for O(1) access, and cheap updates when no changes
+        mapping(bytes32 pmpKeyHash => PMPConfigStorage pmpConfigStorage) pmpConfigsStorage; // slot 1: 32 bytes
+        // config nonce that is incremented during each configureProject call
+        uint8 configNonce; // slot 2: 1 byte
+        // post-configuration hook to be called after a token's PMP is configured
+        IPMPConfigureHook tokenPMPPostConfigHook; // slot 2: 20 bytes
+        // token pmp read augmentation hook to be called when reading a token's PMPs
+        IPMPAugmentHook tokenPMPReadAugmentationHook; // slot 3: 20 bytes
+    }
+
+    // mapping of ProjectConfig structs for each project
+    mapping(address coreContract => mapping(uint256 projectId => ProjectConfig projectConfig)) projectConfigs;
+
+    // mapping of PMP structs for each token
+    mapping(address coreContract => mapping(uint256 tokenId => mapping(bytes32 pmpKeyHash => PMPStorage pmp))) tokenPMPs;
+
+    /**
+     * @notice Configure project hooks for post-configuration and read augmentation.
+     * @param coreContract The address of the core contract.
+     * @param projectId The project ID to configure hooks for.
+     * @param tokenPMPPostConfigHook The hook to call after a token's PMP is configured.
+     * @param tokenPMPReadAugmentationHook The hook to call when reading a token's PMPs.
+     * @dev Only the project artist can configure project hooks.
+     * @dev Both hooks are validated for ERC165 interface compatibility.
+     */
+    function configureProjectHooks(
+        address coreContract,
+        uint256 projectId,
+        IPMPConfigureHook tokenPMPPostConfigHook,
+        IPMPAugmentHook tokenPMPReadAugmentationHook
+    ) external {
+        // only artists may configure project hooks
+        _onlyArtist({
+            coreContract: coreContract,
+            projectId: projectId,
+            sender: msg.sender
+        });
+        // validation - check for ERC165 implementation for non-null hooks
+        if (address(tokenPMPPostConfigHook) != address(0)) {
```

Comment:
What would this `tokenPMPPostConfigHook` be used for?

Reply:
it can be used for effects of PMP updates.
one example:
* a PMP named "color" is set to "red"
* a post-config hook tallies all PMP edits, and reverts if the total number of "red" is above some number, across the entire project

another example:
* a post-config hook tallies the total of each option in a 3-option "select" type PMP
* an augmentation hook grabs this tallied value and injects it into the returned PMPs for the token
* the project's script in the generator now has the total PMP tallies for an entire project, in addition to the individual configuration for the token

its very open ended and could get pretty wild 😆

Reply:
makes sense, thanks for the context!